router/core/operation_processor.go

```diff
@@ -919,37 +926,71 @@ func (o *OperationKit) Validate(skipLoader bool) (cacheHit bool, err error) {
-	valid := globalComplexityResult.Depth <= maxQueryDepth
+	globalComplexityResult, rootFieldStats := operation_complexity.CalculateOperationComplexity(operation, definition, &report)
+	cacheResult := ComplexityCacheEntry{
+		Depth:       globalComplexityResult.Depth,
+		TotalFields: globalComplexityResult.NodeCount,
+	}
+	for _, entry := range rootFieldStats {
+		if entry.Alias == "" {
+			cacheResult.RootFields += 1
+		} else {
+			cacheResult.RootFieldAliases += 1
+		}
+	}
 
-	if o.cache != nil && o.cache.queryDepthCache != nil {
-		o.cache.queryDepthCache.Set(o.parsedOperation.ID, globalComplexityResult.Depth, 1)
+	if o.cache != nil && o.cache.complexityCache != nil {
+		o.cache.complexityCache.Set(o.parsedOperation.ID, cacheResult, 1)
 	}
 
-	if !valid {
-		return false, globalComplexityResult.Depth, &httpGraphqlError{
-			message:    fmt.Sprintf("The query depth %d exceeds the max query depth allowed (%d)", globalComplexityResult.Depth, maxQueryDepth),
-			statusCode: http.StatusBadRequest,
+	return false, cacheResult, o.runComplexityComparisons(complexityLimitConfig, cacheResult, isPersisted)
+}
+
+func (o *OperationKit) runComplexityComparisons(complexityLimitConfig *config.ComplexityLimits, cachedComplexity ComplexityCacheEntry, isPersisted bool) error {
+	type comp struct {
+		field        int
+		cachedField  int
+		errorMessage string
+	}
+
+	testComparisons := []comp{}
+	if complexityLimitConfig.Depth != nil && complexityLimitConfig.Depth.ApplyLimit(isPersisted) {
+		testComparisons = append(testComparisons,
+			comp{complexityLimitConfig.Depth.Limit, cachedComplexity.Depth, fmt.Sprintf("The query depth %d exceeds the max query depth allowed (%d)", cachedComplexity.Depth, complexityLimitConfig.Depth.Limit)})
+	}
+	if complexityLimitConfig.TotalFields != nil && complexityLimitConfig.TotalFields.ApplyLimit(isPersisted) {
+		testComparisons = append(testComparisons,
+			comp{complexityLimitConfig.TotalFields.Limit, cachedComplexity.TotalFields, fmt.Sprintf("The query total fields %d exceeds the max total fields allowed (%d)", cachedComplexity.TotalFields, complexityLimitConfig.TotalFields.Limit)})
+	}
+	if complexityLimitConfig.RootFields != nil && complexityLimitConfig.RootFields.ApplyLimit(isPersisted) {
+		testComparisons = append(testComparisons,
+			comp{complexityLimitConfig.RootFields.Limit, cachedComplexity.RootFields, fmt.Sprintf("The number of root fields %d exceeds the root field limit allowed (%d)", cachedComplexity.RootFields, complexityLimitConfig.RootFields.Limit)})
+	}
+	if complexityLimitConfig.RootFieldAliases != nil && complexityLimitConfig.RootFieldAliases.ApplyLimit(isPersisted) {
+		testComparisons = append(testComparisons,
+			comp{complexityLimitConfig.RootFieldAliases.Limit, cachedComplexity.RootFieldAliases, fmt.Sprintf("The number of root fields aliases %d exceeds the root field aliases limit allowed (%d)", cachedComplexity.RootFieldAliases, complexityLimitConfig.RootFieldAliases.Limit)})
```

Comment:
The number of root field aliases %d exceeds the root field aliases limit allowed (%d)